scripts/extensions/sams/src/store/sets/actions.ts

```diff
@@ -0,0 +1,145 @@
+    };
+}
+
+export function closeSetContentPanel(): ISetActionTypes {
+    return {
+        type: MANAGE_SETS_CLOSE_CONTENT_PANEL,
+    };
+}
+
+export function onManageSetsModalClosed(): ISetActionTypes {
+    return {
+        type: MANAGE_SETS_RESET,
+    };
+}
+
+export function loadSets(): IThunkAction<Array<ISetItem>> {
+    return (dispatch, _getState, {api}) => {
+        return api.sets.getAll()
+            .then((sets: Array<ISetItem>) => {
+                dispatch(receiveSets(sets));
+
+                return Promise.resolve(sets);
+            });
+    };
+}
+
+function openDeleteConfirmationModal(superdesk: ISuperdesk, set: ISetItem): Promise<boolean> {
+    const {gettext} = superdesk.localization;
+    const {confirm} = superdesk.ui;
+
+    const el = document.createElement('div');
+
+    // FIXME: Add an extra backdrop that will cover the Manage Sets modal
+    // This is required because the ui-framework calculates z-index
+    // based on the number of active modals, where as we're using
+    // a mixture of the ui-framework and pure React modals
+    // (superdesk.ui.showModal vs superdesk.ui.confirm)
+    el.classList.add('modal__backdrop', 'fade', 'in');
+    el.style.zIndex = '1050';
+    document.body.append(el);
+
+    return confirm(
+        gettext('Are you sure you want to delete the Set "{{name}}"?', {name: set.name}),
+        gettext('Delete Set?'),
+    )
+        .then((response: boolean) => {
+            el.remove();
+            return response;
+        });
+}
+
+export function confirmBeforeDeletingSet(set: ISetItem): IThunkAction<void> {
+    return (dispatch, getState, {superdesk, api}) => {
+        return openDeleteConfirmationModal(superdesk, set)
+            .then((response: boolean) => {
+                if (response === true) {
+                    return api.sets.delete(set)
+                        .then(() => {
+                            dispatch(removeSetInStore(set));
+
+                            if (getSelectedSetId(getState()) === set._id) {
+                                dispatch(closeSetContentPanel());
+                            }
+                        });
+                }
+
+                return Promise.resolve();
+            });
+    };
+}
+
+export function updateSet(original: ISetItem, updates: Partial<ISetItem>): IThunkAction<ISetItem> {
+    return (dispatch, _getState, {api}) => {
+        return api.sets.update(original, updates)
+            .then((updatedSet: ISetItem) => {
+                dispatch(updatedSetInStore(updatedSet));
+                dispatch(previewSet(updatedSet._id));
+
+                return updatedSet;
+            });
+    };
+}
+
+export function createSet(item: Partial<ISetItem>): IThunkAction<ISetItem> {
+    return (dispatch, _getState, {api}) => {
+        return api.sets.create(item)
+            .then((newSet: ISetItem) => {
+                dispatch(loadSets());
+                dispatch(previewSet(newSet._id));
```

Comment:
Shouldn't preview action be dispatched only after `loadSets` promise is resolved?

Reply:
Probably, but through the use of Redux the preview get's rendered properly anyway. The code works fine

Reply:
It will work fine in most cases when the internet connection is stable. If it's not and HTTP request to fetch sets takes longer, the preview will get opened before the sets are loaded.

We should ensure that there aren't crashes or bugs because of a slower connection. Can you fix this and similar situations and not dispatch actions synchronously that depend on an asynchronous result?

Reply:
https://github.com/superdesk/superdesk-client-core/blob/74e4f93e37ff21e78b6624b4e0e98626fa26958f/scripts/extensions/sams/src/components/sets/setPreviewPanel.tsx#L46

Reply:
It might not crash in this scenario, but it's always better to handle things correctly as close to the top level as possible. For example, if the preview component can't be rendered if the set is null, why even allow nullable set prop instead of ensuring that it's not null before the component is rendered?

Reply:
Updated in https://github.com/superdesk/superdesk-client-core/pull/3552/commits/61cafacbf209090496b108c5132368421d774469